Produce 8 bit @font-face URL strings when possible

heycam opened the pull request:
#### 7995dcceaf8679a715336b9ee9ee4de7793976ac
<pre>
Produce 8 bit @font-face URL strings when possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=252781">https://bugs.webkit.org/show_bug.cgi?id=252781</a>
&lt;rdar://problem/105801659&gt;

Reviewed by Simon Fraser.

We can end up wasting memory storing @font-face URLs (which can be large data:
URLs) as 16 bit strings, if they appear in style sheet text stored as a 16 bit
string, even when the URL itself is 8 bit only.

This uses isAllASCII(), when really we could be checking for &quot;is all
Latin-1&quot;, but there&apos;s an existing fast path for ASCII-checking.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):

Canonical link: <a href="https://commits.webkit.org/260716@main">https://commits.webkit.org/260716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88b4eebc205ad40d0c2b81ee795d83db2673e457

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18220 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41958 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/671 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118387 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113021 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9522 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101380 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114896 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97988 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42928 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29642 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84659 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11000 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30985 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7918 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17101 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50588 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7394 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13347 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -211,7 +211,9 @@ namespace CSSPropertyParserHelpersWorkerSafe {
 {
-    auto location = context.completeURL(CSSPropertyParserHelpers::consumeURLRaw(range).toString());
+    StringView parsedURL = CSSPropertyParserHelpers::consumeURLRaw(range);
+    String url = !parsedURL.is8Bit() && parsedURL.isAllASCII() ? String::make8Bit(parsedURL.characters16(), parsedURL.length()) : parsedURL.toString();
```
Nit: if I see `url` I expect a `URL` type. This is a `urlString`

heycam committed:
Produce 8 bit @font-face URL strings when possible
https://bugs.webkit.org/show_bug.cgi?id=252781
<rdar://problem/105801659>

Reviewed by Simon Fraser.

We can end up wasting memory storing @font-face URLs (which can be large data:
URLs) as 16 bit strings, if they appear in style sheet text stored as a 16 bit
string, even when the URL itself is 8 bit only.

This uses isAllASCII(), when really we could be checking for "is all
Latin-1", but there's an existing fast path for ASCII-checking.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):

Canonical link: https://commits.webkit.org/260716@main
